File: backend/app/services/diagnostics.py
from __future__ import annotations

import os
import platform
import shutil
import subprocess
import sys
from datetime import UTC, datetime
from pathlib import Path

from sqlalchemy import text
from sqlalchemy.engine import Engine

from app.adapters.hardware import GpuSnapshot, HardwareProbe, RamSnapshot
from app.adapters.llm.types import RuntimeHealth
from app.adapters.milvus.types import MilvusHealth
from app.schemas.diagnostics import CheckStatus, DiagnosticCheck, DiagnosticsResponse
from app.settings import Settings


def _display_path(path: Path) -> str:
    try:
        return path.resolve().relative_to(Path.cwd().resolve()).as_posix()
    except ValueError:
        return path.name


def _worst(*statuses: CheckStatus) -> CheckStatus:
    if "error" in statuses:
        return "error"
    if "warning" in statuses:
        return "warning"
    return "ok"


def _python_check() -> DiagnosticCheck:
    version = platform.python_version()
    status: CheckStatus = "ok" if sys.version_info >= (3, 12) else "error"
    return DiagnosticCheck(
        id="python",
        label="Python",
        status=status,
        summary=f"Python {version} ({platform.system()})",
        hint=None if status == "ok" else "Install Python 3.12+.",
    )


def _node_check() -> DiagnosticCheck:
    node = shutil.which("node")
    if node is None:
        return DiagnosticCheck(
            id="node",
            label="Node.js",
            status="warning",
            summary="node is not on PATH.",
            hint="Install Node.js LTS to run the Vite frontend.",
        )
    result = subprocess.run(
        ["node", "--version"], capture_output=True, text=True, timeout=5, check=False
    )
    version = (result.stdout or result.stderr).strip()
    return DiagnosticCheck(
        id="node",
        label="Node.js",
        status="ok" if result.returncode == 0 else "warning",
        summary=version or "node found",
    )


def _data_dir_check(settings: Settings) -> DiagnosticCheck:
    missing = [
        name
        for name, path in {
            "data": settings.data_dir,
            "logs": settings.logs_dir,
            "novels": settings.novels_dir,
        }.items()
        if not path.is_dir()
    ]
    if missing:
        return DiagnosticCheck(
            id="data_dir",
            label="Data directories",
            status="error",
            summary=f"Missing directories: {', '.join(missing)}",
            hint="Restart the backend so it can create DATA_DIR, logs, and novels.",
        )
    return DiagnosticCheck(
        id="data_dir",
        label="Data directories",
        status="ok",
        summary=f"Using {_display_path(settings.data_dir)} (logs/novels/cache present)",
    )


def _sqlite_check(settings: Settings, engine: Engine) -> DiagnosticCheck:
    sqlite_path = settings.sqlite_path
    if sqlite_path is None or not sqlite_path.is_file():
        return DiagnosticCheck(
            id="sqlite",
            label="SQLite",
            status="error",
            summary="SQLite file is missing.",
            hint="Restart the backend to run Alembic migrations.",
        )
    try:
        with engine.connect() as connection:
            version = connection.execute(text("SELECT version_num FROM alembic_version")).scalar()
    except Exception as exc:
        return DiagnosticCheck(
            id="sqlite",
            label="SQLite",
            status="error",
            summary="SQLite is present but schema version could not be read.",
            hint=str(exc.__class__.__name__),
        )
    return DiagnosticCheck(
        id="sqlite",
        label="SQLite",
        status="ok",
        summary=f"{_display_path(sqlite_path)} @ {version}",
    )


def _ollama_check(health: RuntimeHealth) -> DiagnosticCheck:
    if not health.reachable:
        return DiagnosticCheck(
            id="ollama",
            label="Ollama",
            status="error",
            summary=health.message or "Ollama is not reachable.",
            hint="Start Ollama locally. Do not pull large models from CPU_DEV.",
        )
    if not health.default_model_installed:
        return DiagnosticCheck(
            id="ollama",
            label="Ollama",
            status="warning",
            summary=health.message or f"{health.default_model} is not installed.",
            hint="Pull the configured writer model only on the Windows GPU machine.",
        )
    return DiagnosticCheck(
        id="ollama",
        label="Ollama",
        status="ok",
        summary=f"Ollama {health.version or 'ok'}; {health.default_model} installed",
    )


def _milvus_check(health: MilvusHealth) -> DiagnosticCheck:
    status: CheckStatus = "ok" if health.reachable else "warning"
    return DiagnosticCheck(
        id="milvus",
        label="Milvus",
        status=status,
        summary=health.message,
        hint=health.hint or None,
    )


def _ram_check(snapshot: RamSnapshot) -> DiagnosticCheck:
    if not snapshot.ok:
        return DiagnosticCheck(
            id="ram",
            label="RAM",
            status="warning",
            summary=snapshot.message or "RAM probe failed.",
            hint="Diagnostics continues without RAM details.",
            code="ram_unavailable",
        )
    return DiagnosticCheck(
        id="ram",
        label="RAM",
        status="ok",
        summary=snapshot.message,
        code="ram_ok",
    )


def _gpu_check(snapshot: GpuSnapshot) -> DiagnosticCheck:
    if snapshot.presence == "absent":
        return DiagnosticCheck(
            id="gpu",
            label="GPU",
            status="warning",
            summary=snapshot.message,
            hint="Expected on CPU-only hosts. This is not a probe failure.",
            code="gpu_absent",
        )
    if snapshot.presence == "probe_failed":
        return DiagnosticCheck(
            id="gpu",
            label="GPU",
            status="warning",
            summary=snapshot.message,
            hint="GPU detection failed; other diagnostics are unchanged.",
            code="gpu_probe_failed",
        )
    return DiagnosticCheck(
        id="gpu",
        label="GPU",
        status="ok",
        summary=snapshot.message,
        hint="Informational only. Not a VRAM budget or scheduler.",
        code="gpu_ok",
    )


def build_copy_summary(checks: list[DiagnosticCheck], overall: CheckStatus) -> str:
    lines = [
        "Local Novel Studio diagnostics",
        f"overall={overall}",
        f"pid={os.getpid()}",
    ]
    for check in checks:
        extra = f" [{check.code}]" if check.code else ""
        lines.append(f"{check.id}: {check.status}{extra} — {check.summary}")
        if check.hint:
            lines.append(f"  hint: {check.hint}")
    return "\n".join(lines)


async def collect_diagnostics(
    *,
    settings: Settings,
    engine: Engine,
    ollama_health: RuntimeHealth,
    milvus_health: MilvusHealth,
    hardware: HardwareProbe | None = None,
) -> DiagnosticsResponse:
    probe = hardware or HardwareProbe()
    checks = [
        _python_check(),
        _node_check(),
        _data_dir_check(settings),
        _sqlite_check(settings, engine),
        _ollama_check(ollama_health),
        _milvus_check(milvus_health),
        _ram_check(probe.ram()),
        _gpu_check(probe.gpu()),
    ]
    overall = _worst(*(check.status for check in checks))
    return DiagnosticsResponse(
        generated_at=datetime.now(UTC),
        overall_status=overall,
        checks=checks,
        copy_summary=build_copy_summary(checks, overall),
    )

File: frontend/src/lib/diagnostics.ts
export type CheckStatus = "ok" | "warning" | "error";
export type GpuPresence = "ok" | "absent" | "probe_failed" | "unknown";

export function overallLabel(status: CheckStatus): string {
  if (status === "ok") return "正常";
  if (status === "warning") return "警告";
  return "异常";
}

export function statusClassName(status: CheckStatus): string {
  if (status === "ok")
    return "border-emerald-200 bg-emerald-50 text-emerald-900";
  if (status === "warning")
    return "border-amber-200 bg-amber-50 text-amber-900";
  return "border-red-200 bg-red-50 text-red-900";
}

export function gpuPresenceFromCode(
  code: string | null | undefined,
): GpuPresence {
  if (code === "gpu_ok") return "ok";
  if (code === "gpu_absent") return "absent";
  if (code === "gpu_probe_failed") return "probe_failed";
  return "unknown";
}

export function gpuPresenceLabel(kind: GpuPresence): string {
  if (kind === "ok") return "已检测到 GPU";
  if (kind === "absent") return "无 NVIDIA GPU";
  if (kind === "probe_failed") return "GPU 探测失败";
  return "GPU 状态未知";
}

export function summaryLooksPrivate(
  copySummary: string,
  novelBody: string,
): boolean {
  return !copySummary.includes(novelBody);
}

export const CORE_CHECK_IDS = [
  "python",
  "node",
  "data_dir",
  "sqlite",
  "ollama",
  "milvus",
  "ram",
  "gpu",
] as const;

export type CoreCheckId = (typeof CORE_CHECK_IDS)[number];

export const CORE_CHECK_LABELS: Record<CoreCheckId, string> = {
  python: "Python",
  node: "Node.js",
  data_dir: "数据目录",
  sqlite: "SQLite",
  ollama: "Ollama",
  milvus: "Milvus",
  ram: "RAM",
  gpu: "GPU",
};

export type DiagnosticCheck = {
  id: string;
  label: string;
  status: CheckStatus;
  summary: string;
  hint: string | null;
  code?: string | null;
};

export type DiagnosticsResponse = {
  generated_at: string;
  overall_status: CheckStatus;
  checks: DiagnosticCheck[];
  copy_summary: string;
};

const LOADING_ERROR =
  "无法读取诊断接口。请先运行 scripts/dev-backend.ps1，然后刷新。";

export type DiagnosticsView = {
  phase: "loading" | "error" | "ready";
  data: DiagnosticsResponse | null;
  error: string | null;
  checks: DiagnosticCheck[];
};

export function isDiagnosticsResponse(
  value: unknown,
): value is DiagnosticsResponse {
  if (!value || typeof value !== "object") return false;
  const payload = value as DiagnosticsResponse;
  return (
    typeof payload.generated_at === "string" &&
    (payload.overall_status === "ok" ||
      payload.overall_status === "warning" ||
      payload.overall_status === "error") &&
    Array.isArray(payload.checks) &&
    typeof payload.copy_summary === "string"
  );
}

export function withPartialChecks(
  checks: DiagnosticCheck[],
): DiagnosticCheck[] {
  const byId = new Map(checks.map((check) => [check.id, check]));
  return CORE_CHECK_IDS.map((id) => {
    const existing = byId.get(id);
    if (existing) return existing;
    return {
      id,
      label: CORE_CHECK_LABELS[id],
      status: "warning",
      summary: "该项未从诊断接口返回。",
      hint: "刷新诊断。其它检查项仍可查看。",
      code: "partial_missing",
    };
  });
}

export function diagnosticsView(args: {
  previous: DiagnosticsResponse | null;
  incoming:
    | { type: "loading" }
    | { type: "error" }
    | { type: "ok"; payload: unknown };
}): DiagnosticsView {
  if (args.incoming.type === "loading") {
    return {
      phase: args.previous ? "ready" : "loading",
      data: args.previous,
      error: null,
      checks: args.previous ? withPartialChecks(args.previous.checks) : [],
    };
  }
  if (args.incoming.type === "error") {
    return {
      phase: "error",
      data: args.previous,
      error: LOADING_ERROR,
      checks: args.previous ? withPartialChecks(args.previous.checks) : [],
    };
  }
  if (!isDiagnosticsResponse(args.incoming.payload)) {
    return {
      phase: "error",
      data: args.previous,
      error: LOADING_ERROR,
      checks: args.previous ? withPartialChecks(args.previous.checks) : [],
    };
  }
  return {
    phase: "ready",
    data: args.incoming.payload,
    error: null,
    checks: withPartialChecks(args.incoming.payload.checks),
  };
}

export function checkExtraLabel(check: DiagnosticCheck): string | null {
  if (check.id !== "gpu") return null;
  const presence = gpuPresenceFromCode(check.code);
  if (presence === "unknown") return null;
  return gpuPresenceLabel(presence);
}
